### ISSUE TITLE
Aerial Companion displays blank screen when running Aerial on a specific monitor

### ISSUE THREAD

**Issue by yohannes** [AerialScreensaver/Aerial#19]
Running v1.2 with Aerial 2.2.1 on MacOS 11.0.1.

I notice that AC takes up memory over time (over 1GB in average) presumably as it downloads / caches contents but nothing is being shown on screen. It's just blank / dark. 

Best,
Yohannes

**Comment by glouel** [comment]
Hey @yohannes 

Just to clarify, that's after launching "Experimental > Launch Aerial Fullscreen on this monitor", correct ? And with the window (fullscreen or minimised) still up ? How long did you wait approximately ? 

One issue with that feature right now is that when Companion runs Aerial, it doesn't load the settings/cache that's used on Big Sur by the "regular" screensaver (because of containerisation of the "regular" screensaver, Companion is running the same file, but it "runs" outside the container in a different path, just like it used to on Mojave and earlier). 

So it will try to download a fresh video which may take some time and could explain what you are seeing. I'll keep an eye out but try closing and launching again a couple of times, once you get one video usually things are much better. 

Another option I'm working on is copying the settings from the container to the normal place, and changing the cache path automatically for those copied settings to the one in the container. This may take me a bit to get right though but I'm on it.

**Comment by yohannes** [comment]
> Just to clarify, that's after launching "Experimental > Launch Aerial Fullscreen on this monitor", correct ? And with the window (fullscreen or minimised) still up ? How long did you wait approximately ?

Yes, correct. Both fullscreen and not fullscreen. Approximately 5 mins. 

Well noted on your additional remarks especially in regards to the cache inaccessibility imposed by the containerization policy in Big Sur. That makes sense for the particular case here. Thanks.

**Comment by glouel** [comment]
Thanks @yohannes 

So, a few thoughts that may help troubleshoot this.

- First, if you were using Aerial in Mojave, you may still have an old configuration file with something that the new Aerial can't quite cope with. A custom path could be one, though there are other things that may cause an issue. So, you can delete those preferences by doing this in terminal : 

`defaults -currentHost delete com.JohnCoates.Aerial`

Note that if you are unfamiliar with Terminal, the file is in `~/Library/Preferences/ByHost`. But if you delete it by hand, because macOS caches those, you'll have to reboot your mac so it gets fully cleared up.

- Second, the log file from the `~/Library/Application Support/Aerial` may tell me more about what's happening, if you can attach it here. It would be interesting also to check if you have the various tvOS folders or not in there.

- Third, if you happen to use Little Snitch, you'll need to create new rules for Aerial Companion as it's the one running Aerial now, and not `legacyScreenSaver.appex`. Though in that case you may have been prompted to do so already (I'm not sure). 

Let me know how this goes, thanks !

**Comment by glouel** [comment]
So just to let you know, I updated both Aerial and Aerial Companion. 

Latest version of companion is here, in beta : https://github.com/glouel/AerialCompanion/releases/download/v1.2.1beta1/AerialCompanion.dmg

I also posted a standalone version of Aerial as an App, that lets you change settings (a workaround until I fix the issue and integrate it in Companion) : https://github.com/glouel/AerialCompanion/releases/download/v1.2.1beta1/AerialApp.app.zip

I did find a "memory leak" of sorts that's fixed by Aerial 2.2.2beta1, though I doubt that was your issue (it was keeping multiple instances of the video decoder if you launched/closed/launched again the windowed mode). You can get it from Companion by changing desired version to beta.

**Comment by yohannes** [comment]
Hi @glouel 

Thanks for such an informational reply. I did install the 1.2.1beta1 companion app and right off the bat, the feature of launching aerial in a specific monitor worked - the screensaver started playing. Then, I installed the 1.2.1b1 standalone aerial app just for kick & after looking around in it, I uninstalled it since I thought I could change the settings as well in the screensaver version. Right after that, the feature of launching aerial in a specific monitor no longer worked. I thought that's kinda weird cos I didn't change any setting . All I did was install the standalone app. So, I reinstalled the companion app but it didn't work. So I thought I killed the companion app, deleted the existing Aerial Log, relaunched the  companion app and here's the new generated log (below). I went through it and the potential red flag I noticed was "2020-12-03 16:11:54.790 : Empty playlist, not good !".


```
2020-12-03 16:11:54.678 : avInit .saver (0.0, 0.0, 1280.0, 720.0) p: false o: false
2020-12-03 16:11:54.679 : Time Management initialized
2020-12-03 16:11:54.679 : <AerialView: 0x7f8dbe52f0d0> AerialView setup init (V2.2.1) preview: false
2020-12-03 16:11:54.680 : Running Version 11.0.1 (Build 20B29)
2020-12-03 16:11:54.680 : Display Detection initialized
2020-12-03 16:11:54.680 : ***Display Detection***
2020-12-03 16:11:54.680 : 3 display(s) detected
2020-12-03 16:11:54.681 : 1cm = 40.157480918065474 points
2020-12-03 16:11:54.681 : npass: dict [__C.NSDeviceDescriptionKey(_rawValue: NSDeviceColorSpaceName): NSCalibratedRGBColorSpace, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceResolution): NSSize: {72, 72}, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceSize): NSSize: {1920, 1080}, __C.NSDeviceDescriptionKey(_rawValue: NSScreenNumber): 188943304, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceIsScreen): YES, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceBitsPerSample): 8]
2020-12-03 16:11:54.681 :        bottomLeftFrame (0.0, 0.0, 1920.0, 1080.0)
2020-12-03 16:11:54.681 : npass: dict [__C.NSDeviceDescriptionKey(_rawValue: NSDeviceColorSpaceName): NSCalibratedRGBColorSpace, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceResolution): NSSize: {72, 72}, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceSize): NSSize: {1920, 1080}, __C.NSDeviceDescriptionKey(_rawValue: NSScreenNumber): 188943303, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceIsScreen): YES, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceBitsPerSample): 8]
2020-12-03 16:11:54.681 :        bottomLeftFrame (1920.0, 0.0, 1920.0, 1080.0)
2020-12-03 16:11:54.682 : npass: dict [__C.NSDeviceDescriptionKey(_rawValue: NSDeviceColorSpaceName): NSCalibratedRGBColorSpace, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceResolution): NSSize: {72, 72}, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceSize): NSSize: {1920, 1080}, __C.NSDeviceDescriptionKey(_rawValue: NSScreenNumber): 188943305, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceIsScreen): YES, __C.NSDeviceDescriptionKey(_rawValue: NSDeviceBitsPerSample): 8]
2020-12-03 16:11:54.682 :        bottomLeftFrame (-1920.0, 0.0, 1920.0, 1080.0)
2020-12-03 16:11:54.682 : src orig : (0.0, 0.0)
2020-12-03 16:11:54.682 : left 1.0 below 0.0
2020-12-03 16:11:54.683 : src orig : (1920.0, 0.0)
2020-12-03 16:11:54.683 : left 2.0 below 0.0
2020-12-03 16:11:54.683 : src orig : (-1920.0, 0.0)
2020-12-03 16:11:54.683 : left 0.0 below 0.0
2020-12-03 16:11:54.683 : [id=188943304, width=1920, height=1080, bottomLeftFrame=(0.0, 0.0, 1920.0, 1080.0), topRightCorner=(1920.0, 1080.0), isMain=true, backingScaleFactor=1.0]
2020-12-03 16:11:54.684 : [id=188943303, width=1920, height=1080, bottomLeftFrame=(1920.0, 0.0, 1920.0, 1080.0), topRightCorner=(3840.0, 1080.0), isMain=false, backingScaleFactor=1.0]
2020-12-03 16:11:54.684 : [id=188943305, width=1920, height=1080, bottomLeftFrame=(-1920.0, 0.0, 1920.0, 1080.0), topRightCorner=(0.0, 1080.0), isMain=false, backingScaleFactor=1.0]
2020-12-03 16:11:54.684 : (-1920.0, 0.0, 5760.0, 1080.0)
2020-12-03 16:11:54.684 : ***Display Detection Done***
2020-12-03 16:11:54.685 : Real screen count : 3
2020-12-03 16:11:54.685 : Using : nil
2020-12-03 16:11:54.685 : <AerialView: 0x7f8dbe52f0d0> setting up player layer with bounds/frame: (0.0, 0.0, 1280.0, 720.0) / (0.0, 0.0, 1280.0, 720.0)
2020-12-03 16:11:54.686 : Using dot workaround for video driver corruption
2020-12-03 16:11:54.687 : Last manifest check : 2020-12-02 17:00:00 +0000
2020-12-03 16:11:54.687 : tvOS 13 is enabled, cached and up to date
2020-12-03 16:11:54.687 : Last manifest check : 2020-12-02 17:00:00 +0000
2020-12-03 16:11:54.687 : tvOS 12 is enabled, cached and up to date
2020-12-03 16:11:54.688 : Last manifest check : 2020-12-02 17:00:00 +0000
2020-12-03 16:11:54.688 : tvOS 11 is enabled, cached and up to date
2020-12-03 16:11:54.688 : Last manifest check : 2020-12-02 17:00:00 +0000
2020-12-03 16:11:54.688 : tvOS 10 is enabled, cached and up to date
2020-12-03 16:11:54.689 : Last manifest check : 2020-12-02 17:00:00 +0000
2020-12-03 16:11:54.689 : From Joshua Michaels & Hal Bergman is enabled, cached and up to date
2020-12-03 16:11:54.780 : <AerialView: 0x7f8dbe52f0d0> backing change 1.0 isDisabled: false frame: (0.0, 0.0, 1280.0, 720.0) preview: false
2020-12-03 16:11:54.786 : Refreshing video list
2020-12-03 16:11:54.787 : Poi Strings Provider initialized
2020-12-03 16:11:54.787 : Preferred languages : ["en-ID"]
2020-12-03 16:11:54.787 : Best matched language : en
2020-12-03 16:11:54.787 : Defaulting to bundle
2020-12-03 16:11:54.788 : path : /Users/yohanes.wijaya/Library/Screen Savers/Aerial.saver/Contents/Resources/en.json
2020-12-03 16:11:54.788 : Community JSON : 99 entries
2020-12-03 16:11:54.788 : source : tvOS 13 contains 90 new videos (total 90)
2020-12-03 16:11:54.788 : source : tvOS 12 contains 0 new videos (total 90)
2020-12-03 16:11:54.788 : source : tvOS 11 contains 0 new videos (total 90)
2020-12-03 16:11:54.789 : source : tvOS 10 contains 9 new videos (total 99)
2020-12-03 16:11:54.789 : source : From Joshua Michaels & Hal Bergman contains 20 new videos (total 119)
2020-12-03 16:11:54.789 : <AerialView: 0x7f8dbe52f0d0> pnv
2020-12-03 16:11:54.789 : <AerialView: 0x7f8dbe52f0d0> Setting player for all player layers in []
2020-12-03 16:11:54.790 : remaining in playlist : 0
2020-12-03 16:11:54.790 : generate playlist
2020-12-03 16:11:54.790 : playlist count: 119
2020-12-03 16:11:54.790 : Final count : 0
2020-12-03 16:11:54.790 : Empty playlist, not good !
2020-12-03 16:11:54.791 : returning random video after condition change not met !
2020-12-03 16:11:54.791 : initvideocache
2020-12-03 16:11:54.791 : <AerialView: 0x7f8dbe52f0d0> streaming video (not fully available offline) : https://sylvan.apple.com/Videos/g201_AK_A003_C014_SDR_20191113_SDR_2K_AVC.mov
2020-12-03 16:11:55.492 : Processsing contentInformationRequest
2020-12-03 16:11:55.493 : expected content length: 498623146 type:com.apple.quicktime-movie
2020-12-03 16:11:55.502 : Processsing contentInformationRequest
2020-12-03 16:11:55.502 : expected content length: 498623146 type:com.apple.quicktime-movie
2020-12-03 16:11:56.661 : connectionDidFinishLoading

```

**Comment by glouel** [comment]
Hey @yohannes 

Thanks for your detailed explanation of your steps, I think I know what's happening.

> I did install the 1.2.1beta1 companion app and right off the bat, the feature of launching aerial in a specific monitor worked - the screensaver started playing. 

Ok so my guess was, during the black screen time you experienced on the previous launches, Aerial managed to cache at least one video (more on that below). 

> Then, I installed the 1.2.1b1 standalone aerial app just for kick & after looking around in it, I uninstalled it since I thought I could change the settings as well in the screensaver version.

So by doing that, you probably went through the initial setup thing ? It's possible at this point that some setting got changed (like the resolution), either by you or by Aerial trying to be helpful. And your cache got wiped here. 

> Right after that, the feature of launching aerial in a specific monitor no longer worked. I thought that's kinda weird cos I didn't change any setting . All I did was install the standalone app. So, I reinstalled the companion app but it didn't work. So I thought I killed the companion app, deleted the existing Aerial Log, relaunched the companion app and here's the new generated log (below). I went through it and the potential red flag I noticed was "2020-12-03 16:11:54.790 : Empty playlist, not good !".

Yeah so at that point you have an empty cache, and it tries to stream a video and looks like it fails. I don't remember if I asked, do you happen to use Little Snitch or not ? 

I'll have a second look at the streaming code but please let me know if you use Little Snitch (or another firewall) or not.

**Comment by glouel** [comment]
@yohannes I found the bug, it's actually a longstanding bug with 2.0 and streaming, I'll push the fix soon.

**Comment by glouel** [comment]
@yohannes : please update to Aerial 2.2.3 and it will be all fixed. Note that you still need the AerialApp to change settings, the settings for that windowed/desktop mode are separate from the "classic" screensaver settings on Big Sur/Catalina. Thanks again for your detailed steps, this was a bug that was plaguing Aerial for a while now.

**Comment by yohannes** [comment]
Hi @glouel,

Sorry for the late reply. 

> So by doing that, you probably went through the initial setup thing ? It's possible at this point that some setting got changed (like the resolution), either by you or by Aerial trying to be helpful. And your cache got wiped here.

Yes as soon as the standalone aerial app was launched the first time, I was presented w/  the initial setup and I remembered now that I did change the resolution (so my claim before that I didn't change anything was inaccurate) and so indeed understandably the cache got deleted / overwritten. 

> do you happen to use Little Snitch or not ?

I use similar app called Lulu by Objective-See

> please update to Aerial 2.2.3 and it will be all fixed.

Yes it's working now. If I may give a suggestion though is that the first time it runs (on empty cache), I think it's better to have a loading icon or indicator instead of just a blank / black screen as user with a relatively slow internet connection (like in my case) may need, in hindsight, to wait as content is being downloaded instead of being inpatient & think that it doesn't work. 

Thanks again for the support. I'm closing this issue.
